### DISCUSSION TITLE
Deprecate `echo`

### DISCUSSION THREAD

**Discussion by kaathewisegit** [nushell/nushell#12936] [General]
I know there were prior discussions on this, but I wasn't able to find them.  I'd appreciate it if someone links those here.  But I still want to open this as a discussion (instead of an issue), to have a more free-form conversation.

The UNIX `echo` has a print-oriented purpose: it's [description][man-echo] says "display a line of text".  And the options are geared towards editing the string contents: either escapes or newlines.

Contrary to that, Nushell's `echo` deals in types, not printing to the output.  And, in the majority of use cases, it's useless:

- With a single argument Nushell-to-Nushell values, `... | echo $arg | ...` can be replaced with plain `... | $arg | ...`.

- On the border between Nushell and external programs, `echo` doesn't do any conversion.  `echo (ls) | head` is equivalent to `ls | head`, since it's the pipelines which convert Nu values to string.

The only non-redundant use case is with rest argument, where `echo` converts them to a list.  I'd argue this is an unintuitive feature, which is better dedicated to a differently named command[^list].

To sum up, Nushell `echo`, despite having the same name as the UNIX one, doesn't have the same functionality.  I'd wager a guess that this does more to confuse the newcomers from another shells, than the benefits it might provide.  As such, I'd suggest deprecating and removing it.

[man-echo]: https://linux.die.net/man/1/echo

[^list]: For example, `list (...rest)`, or even `$val | into list (...rest)`.

**Comment by fdncred** [comment]
This is a tricky case, which is why we still have it. Nushell creates structured output so while people compare nushell to bash, it's not really a good comparison since bash doesn't create structured output, but nushell is a shell, of course, so it's in the same arena but does things a little different. Of course, you know all this.

So, echo exists to put structured data, which we call nushell values, on/in the pipeline. You're right that using a variable by itself does the same thing. However, without echo, people will ask for it since there's so much muscle memory there. The print command works more like bash's echo and just prints dumb unstructured text, therefore it's unavailable to be used with piping commands since pipes are looking for nushell values (probably not a 100% correct statement).

I don't think echo converts rest arguments to a list, rest arguments are by their design, a list.

So, I understand the confusion but I'm not sure removing echo would increase or decrease it. I'm conflicted.

**Reply by kaathewisegit** [reply]
> rest arguments are by their design, a list

I've never thought about it this way, but now I realize that this is indeed true.  The only difference comes from how commands interpret it.

On `echo`, the reason I'm leaning towards removal, is because Nushell is sufficiently different to be taken a paradigm shift, in my opinion.  The difference between between structured vs unstructured output, as You've noted.  As I understand it, in POSIX shell `echo` is needed to distinguish between "expand this to a string" vs "expand this to a command and execute it".  But in Nushell, this distinction doesn't exist: it doesn't have the C-preprocessor-like expansion.  So, I see `echo` as a quoting-like device.  One can use echo and `$"($string)"`, but they probably should be steered away from it.

I know that people want `echo`, but I feel like the familiarity it gives is a false one.  [False friends of the translator](https://en.wikipedia.org/wiki/False_friend) kind of deal.
